README.md
[![Build Status](https://travis-ci.org/nalch/aurelia-logging-api.svg?branch=master)](https://travis-ci.org/nalch/aurelia-logging-api)
# `aurelia-logging-api`

> Catch, bundle and send all aurelia logs to an api endpoint.

`aurelia-logging-api` provides an easy to use log appender for Aurelia's LogManager to bundle logs and send them to an endpoint.

The received logs are enriched with the timestamp, source logger and loglevel of the message. The actual message will be sent as list or joined with a configurable separator:
```js
[
  {
    timestamp: '2020-01-01T12:00:00Z',
    logger: 'my-app-logger',
    loglevel: 'info',
    message: ['log', 'data', 'list']
  },
  {
    ...
  }
]
```

It adds the abilities to throttle sending after failed requests and catches the following javascript information/errors if configured to do so:
* window errors
* unhandled promise rejections
* requests/responses to the api

## Installation
```bash
npm i aurelia-logging-api --save
```

## Usage
`aurelia-logging-api` uses `aurelia-fetch-client` or `aurelia-http-client` to send logs to a server by providing it with the application's client 
or creating a new one. This makes reusing the connection configuration and authorization easier.
Use the plugin by loading it in the app's `main.js`. The configuration's default values are shown, but can be changed or omitted.
```js
import { HttpClient } from 'aurelia-fetch-client';

...

config = {
  targetUrl: '',
  client: new HttpClient(),
  bufferSize: 100,
  minLevel: logLevel.debug,
  joinMessage: false,
  requestMethod: 'post',
  requestParams: {},
  throttleRetries: true,
  retryIntervals: [5, 60, 3600],
  catchWindowErrors: false,
  catchPromiseRejections: false,
  flushOnUnload: true
};
aurelia.use.plugin('aurelia-logging-api', config);
// for webpack user, use PLATFORM.moduleName wrapper
aurelia.use.plugin(PLATFORM.moduleName('aurelia-logging-api'), config);
```

## Using the Appender
All used logging appenders are registered with Aurelia's dependency injection under the `BackendAppender` class:
```js
...

import { BackendAppender } from 'aurelia-logging-api';

@inject(BackendAppender)
class App {
  constructor(appender) {
    this.appender = appender;
  }

  irregularSend() {
    this.appender.requestQueue.sendRequest();
  }
}
```

## Configuration

| Configuration Key      | Default       | Description |
| ---------------------- | ------------- | ----------- |
| targetUrl              | ''            | Url to send the logs to. |
| client                 | null          | The http (fetch) client to be used to send the logs. Can be intercepted to log the requests/responses as well. |
| bufferSize             | 100           | Number of log messages to be bundled until the batch is sent to the server. |
| minLevel               | debug (40)    | Min level to include in the collection of logs. Usually used to restrict lower level priority logs to the frontend instead of sending to the api. |
| joinMessage            | false         | If set to a string log message data is joined by the separator and sent as string instead of list. |
| requestMethod          | 'post'        | Request method to use for sending the logs. |
| requestParams          | {}            | Arbitrary request parameters to use for the log sending. Only needed, if they are not set on the provided client. |
| throttleRetries        | true          | Throttle subsequent requests, if a request fails. |
| retryIntervals         | [5, 60, 3600] | Intervals to wait between retrying a failed log request. If more fails occur than intervals are given, the last value is used. After a successfull request all data is retried without waiting. |
| catchWindowErrors      | false         | Catch all javascript window errors and log them as errors. |
| catchPromiseRejections | false         | Catch all unhandled promise rejections and log them as errors. |
| flushOnUnload          | false         | Flush the queue, when the window is reloaded or closed. |

## Additional error catching
`aurelia-logging-api` provides some convenience function to catch and log the following errors in the application.
### Window errors
All javascript errors caught by the window's error handler can be logged as error by setting the config's `catchWindowErrors` value to `true`.
### Undhandled promise rejections
All unhandled promise rejections (if supported by the browser) can be logged as error by setting the config's `catchPromiseRejections` value to `true`.
### Api requests/responses
All requests and responses to and from a client can be logged as info by adding `aurelia-logging-api`'s `apiRequestLogger` interceptor:
```js
import { addRequestLogger } from 'aurelia-logging-api';

...

const client = new HttpClient().configure((config) => {
  config.withInterceptor(apiRequestLogger);
});
```


CHANGELOG.md
## [1.3.5](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.5) (2020-01-01)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))
* update for version 1.3.0 ([0bd30d4](https://github.com/nalch/aurelia-logging-api/commit/0bd30d4))
* Update loglevel handling ([7a0a258](https://github.com/nalch/aurelia-logging-api/commit/7a0a258))
* Update loglevel handling ([4211cbb](https://github.com/nalch/aurelia-logging-api/commit/4211cbb))



## [1.3.4](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.4) (2019-12-02)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))
* update for version 1.3.0 ([0bd30d4](https://github.com/nalch/aurelia-logging-api/commit/0bd30d4))
* Update loglevel handling ([7a0a258](https://github.com/nalch/aurelia-logging-api/commit/7a0a258))
* Update loglevel handling ([4211cbb](https://github.com/nalch/aurelia-logging-api/commit/4211cbb))



## [1.3.3](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.3) (2019-12-02)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))
* update for version 1.3.0 ([0bd30d4](https://github.com/nalch/aurelia-logging-api/commit/0bd30d4))
* Update loglevel handling ([7a0a258](https://github.com/nalch/aurelia-logging-api/commit/7a0a258))
* Update loglevel handling ([4211cbb](https://github.com/nalch/aurelia-logging-api/commit/4211cbb))



## [1.3.2](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.2) (2019-12-01)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* linting update ([dcb4e2a](https://github.com/nalch/aurelia-logging-api/commit/dcb4e2a))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))
* update for version 1.3.0 ([0bd30d4](https://github.com/nalch/aurelia-logging-api/commit/0bd30d4))
* Update loglevel handling ([7a0a258](https://github.com/nalch/aurelia-logging-api/commit/7a0a258))
* Update loglevel handling ([4211cbb](https://github.com/nalch/aurelia-logging-api/commit/4211cbb))



# [1.3.1](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.1) (2019-11-30)

### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))
* update for version 1.3.0 ([0bd30d4](https://github.com/nalch/aurelia-logging-api/commit/0bd30d4))
* Update loglevel handling ([4211cbb](https://github.com/nalch/aurelia-logging-api/commit/4211cbb))



# [1.3.0](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.3.0) (2019-09-21)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))
* Update publishing command ([dc7d51c](https://github.com/nalch/aurelia-logging-api/commit/dc7d51c))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))



# [1.2.0](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.2.0) (2019-09-21)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))
* Register appender with dependency injection ([b114c9b](https://github.com/nalch/aurelia-logging-api/commit/b114c9b))



## [1.1.0](https://github.com/nalch/aurelia-logging-api/compare/v1.0.3...v1.1.0) (2019-09-18)


### Bug Fixes

* ci version update ([3bcbba5](https://github.com/nalch/aurelia-logging-api/commit/3bcbba5))


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))



## [1.0.3](https://github.com/nalch/aurelia-logging-api/compare/v1.0.2...v1.0.3) (2019-09-18)


### Features

* Add compatibility with aurelia-http-client ([83e607c](https://github.com/nalch/aurelia-logging-api/commit/83e607c))



## 1.0.2 (2019-09-11)
- Enable flushing the queue on window unonload

## 1.0.1 (2019-09-08)
- Initial commit


src/request-queue.js
import { inject } from 'aurelia-framework';
import { HttpClient, json } from 'aurelia-fetch-client';
import { HttpClient as DeprecatedClient } from 'aurelia-http-client';

@inject(HttpClient, DeprecatedClient)
export class RequestQueue {
  config = {
    requestMethod: 'post',
    requestParams: {},
    throttleRetries: true,
    retryIntervals: [5, 60, 3600]
  };

  requestQueue = [];

  failedAttempts = 0;
  runningRequest = false;

  constructor(config, client, deprecatedClient) {
    if (config) {
      Object.assign(this.config, config);
    }
    this.client = config.client || client || deprecatedClient;
    if (!this.client) {
      throw new Error(
        'No http client found. Pass the client in the config under the key ' +
        '"client" or register it under aurelia-fetch-client\'s or ' +
        'aurelia-http-client\'s HttpClient class in the DI container.'
      );
    }
  }

  push(requestData) {
    const payload = Object.assign(this.config.requestParams, {
      method: this.config.requestMethod,
      body: json(requestData)
    });
    this.requestQueue.push(payload);

    if (!this.runningRequest) {
      this.sendRequest();
    }
  }

  getThrottleInterval() {
    let waitInterval = 0;
    if (this.config.throttleRetries) {
      if (this.failedAttempts < this.config.retryIntervals.length) {
        waitInterval = this.config.retryIntervals[this.failedAttempts];
      } else {
        waitInterval = this.config.retryIntervals[
          this.config.retryIntervals.length - 1
        ];
      }
    }
    return waitInterval * 1000;
  }

  rescheduleRequest() {
    window.setTimeout(
      this.sendRequest.bind(this),
      this.getThrottleInterval()
    );
  }

  sendWithClient(url, payload) {
    if (this.client instanceof HttpClient) {
      return this.client.fetch(this.config.targetUrl, payload);
    } else if (this.client instanceof DeprecatedClient) {
      const clientMethod = this.client[
        this.config.requestMethod].bind(this.client);
      return clientMethod(this.config.targetUrl, payload);
    }
  }

  sendRequest() {
    if (this.requestQueue.length === 0) {
      return;
    }

    this.runningRequest = true;
    const payload = this.requestQueue[0];
    this.sendWithClient(this.config.targetUrl, payload)
      .then((response) => {
        if (!response.ok) {
          throw Error(response.status);
        }
        this.requestQueue.splice(0, 1);
        this.failedAttempts = 0;
      })
      .catch((error) => {
        console.error(`Couldn't send logs to endpoint: ${error}`);  // eslint-disable-line no-console
        this.failedAttempts += 1;
      })
      .finally(() => {
        if (this.requestQueue.length > 0) {
          this.rescheduleRequest();
        } else {
          this.runningRequest = false;
        }
      });
  }
}
